swarm/bmt/bmt.go

```diff
@@ -244,130 +257,148 @@ func (t *Tree) Draw(hash []byte, d int) string {
 	return strings.Join(rows, "\n") + "\n"
 }
 
-// NewTree initialises the Tree by building up the nodes of a BMT
-// segment size is stipulated to be the size of the hash
-// segmentCount needs to be positive integer and does not need to be
-// a power of two and can even be an odd number
-// segmentSize * segmentCount determines the maximum chunk size
-// hashed using the tree
-func NewTree(hasher BaseHasher, segmentSize, segmentCount int) *Tree {
-	n := NewNode(0, 0, nil)
-	n.root = true
-	prevlevel := []*Node{n}
-	// iterate over levels and creates 2^level nodes
-	level := 1
+// newTree initialises a tree by building up the nodes of a BMT
+// - segment size is stipulated to be the size of the hash
+func newTree(segmentSize, depth int) *tree {
+	n := newNode(0, nil)
+	prevlevel := []*node{n}
+	// iterate over levels and creates 2^(depth-level) nodes
 	count := 2
-	for d := 1; d <= depth(segmentCount); d++ {
-		nodes := make([]*Node, count)
-		for i := 0; i < len(nodes); i++ {
+	for level := depth - 2; level >= 0; level-- {
+		nodes := make([]*node, count)
+		for i := 0; i < count; i++ {
 			parent := prevlevel[i/2]
-			t := NewNode(level, i, parent)
-			nodes[i] = t
+			nodes[i] = newNode(i, parent)
 		}
 		prevlevel = nodes
-		level++
 		count *= 2
 	}
-	// the datanode level is the nodes on the last level where
-	return &Tree{
-		leaves: prevlevel,
+	// the datanode level is the nodes on the last level
+	return &tree{
+		leaves:  prevlevel,
+		result:  make(chan []byte, 1),
+		segment: make([]byte, segmentSize),
+		section: make([]byte, 2*segmentSize),
 	}
 }
 
 // methods needed by hash.Hash
 
 // Size returns the size
 func (h *Hasher) Size() int {
-	return h.size
+	return h.pool.SegmentSize
 }
 
 // BlockSize returns the block size
 func (h *Hasher) BlockSize() int {
-	return h.blocksize
+	return h.pool.SegmentSize
+}
+
+// Hash hashes the data and the span using the bmt hasher
+func Hash(h *Hasher, span, data []byte) []byte {
+	h.ResetWithLength(span)
+	h.Write(data)
+	return h.Sum(nil)
+}
+
+// Datalength returns the maximum data size that is hashed by the hasher =
+// segment count times segment size
+func (h *Hasher) DataLength() int {
+	return h.pool.Datalength
 }
 
 // Sum returns the hash of the buffer
 // hash.Hash interface Sum method appends the byte slice to the underlying
 // data before it calculates and returns the hash of the chunk
+// caller must make sure Sum is not called concurrently with Write or WriteSegment
 func (h *Hasher) Sum(b []byte) (r []byte) {
+	return h.sum(b, true, true)
+}
+
+// sum implements Sum taking parameters
+// * if the tree is released right away
+// * if sequential write is used (can read sections)
+func (h *Hasher) sum(b []byte, release, section bool) (r []byte) {
```

Comment:
Are the release and section parameters are for the future? Because it is a private function which is only called with true, true, do we really need these?

Reply:
yes indeed, no release is for when we get proof or verify proof, no section is for when we use the SegmentWriter interface for intermediate chunks for a future version of the chunker split